src/modules/dwm.cpp

```diff
@@ -0,0 +1,493 @@
+    const auto* next = &layout + 1;
+    const auto* first = m_layouts->begin().base();
+    if (next < m_layouts->end().base()) {
+      return next;
+    } else if (wrap) {
+      return first;
+    } else {
+      return &layout;
+    }
+  }
+
+  auto dwm_module::prev_layout(const dwmipc::Layout& layout, bool wrap) const -> const dwmipc::Layout* {
+    const auto* prev = &layout - 1;
+    const auto* last = m_layouts->end().base() - 1;
+    if (prev >= m_layouts->begin().base()) {
+      return prev;
+    } else if (wrap) {
+      return last;
+    } else {
+      return &layout;
+    }
+  }
+
+  void dwm_module::update_tag_labels() {
+    for (auto& t : m_tags) {
+      t.state = get_state(t.bit_mask);
+      t.label = m_state_labels.at(t.state)->clone();
+      t.label->reset_tokens();
+      t.label->replace_token("%name%", t.name);
+    }
+  }
+
+  void dwm_module::update_title_label(const string& title) {
+    m_title_label->reset_tokens();
+    m_title_label->replace_token("%title%", title);
+  }
+
+  void dwm_module::update_title_label() {
+    std::string new_title;
+    if (m_focused_client_id != 0) {
+      try {
+        new_title = m_ipc->get_client(m_focused_client_id)->name;
+      } catch (const dwmipc::SocketClosedError& err) {
+        m_log.err("%s: Disconnected from socket: %s", name(), err.what());
+        sleep(chrono::duration<double>(1));
+        reconnect_dwm();
+      } catch (const dwmipc::IPCError& err) {
+        throw module_error(err.what());
+      }
+    }
+    update_title_label(new_title);
+  }
+
+  void dwm_module::update_floating_label() {
+    if (m_focused_client_id != 0) {
+      try {
+        m_is_floating = m_ipc->get_client(m_focused_client_id)->states.is_floating;
+      } catch (const dwmipc::SocketClosedError& err) {
+        m_log.err("%s: Disconnected from socket: %s", name(), err.what());
+        sleep(chrono::duration<double>(1));
+        reconnect_dwm();
+      } catch (const dwmipc::IPCError& err) {
+        throw module_error(err.what());
+      }
+    } else {
+      m_is_floating = false;
+    }
+  }
+
+  void dwm_module::update_layout_label(const string& symbol) {
+    m_layout_label->reset_tokens();
+    m_layout_label->replace_token("%symbol%", symbol);
+  }
+
+  auto dwm_module::reconnect_dwm() -> bool {
+    try {
+      if (!m_ipc->is_main_socket_connected()) {
+        m_log.notice("%s: Attempting to reconnect to main socket", name());
+        m_ipc->connect_main_socket();
+        m_log.notice("%s: Successfully reconnected to main socket", name());
+      }
+      if (!m_ipc->is_event_socket_connected()) {
+        m_log.notice("%s: Attempting to reconnect event socket", name());
+        m_ipc->connect_event_socket();
+        m_log.notice("%s: Successfully reconnected event to socket", name());
+      }
+      return true;
+    } catch (const dwmipc::IPCError& err) {
+      m_log.err("%s: Failed to reconnect to socket: %s", name(), err.what());
```

Comment:
What happens if this happens?

Reply:
There should probably be a throw module_error there.

Reply:
I agree

Reply:
See 30052f4549495abd320f1f5bb037320ab3fbfcc4